veg2hab/mozaiek.py

```diff
@@ -93,76 +104,301 @@ def __str__(self):
+    ):
+        # Casten naar list voor (de)serializatie
+        self.tegengekomen_kwal_SBB = list(
+            {vegtype for vegtype in tegengekomen_vegtypen if isinstance(vegtype, SBB)}
         )
+        self.tegengekomen_kwal_VvN = list(
+            {vegtype for vegtype in tegengekomen_vegtypen if isinstance(vegtype, VvN)}
+        )
+
+    def _bepaal_kwalificerende_en_HXXXX_omringing(
+        self, omringd_door: pd.DataFrame
+    ) -> Tuple[Union[int, float], Union[int, float], Set[Union[SBB, VvN]]]:
+        omringing_kwal_vlakken = 0
+        omringing_HXXXX = 0
+        tegengekomen_kwal_vegtypen = set()
+
+        grouped_by_vlak = omringd_door.groupby("ElmID")
+
+        for _, vlak_group in grouped_by_vlak:
+            assert (
+                len(vlak_group.omringing_percentage.unique()) == 1
+            ), f"Omringing percentage is niet hetzelfde voor alle complexdelen in vlak met ElmID {vlak_group.ElmID.iloc[0]}"
+
+            (
+                bedekking_kwal_complexdelen,
+                bedekking_HXXXX,
+                tegengekomen_kwal_vegtypen_vlak,
+            ) = self._bepaal_kwalificerende_en_HXXXX_bedekking(vlak_group)
+
+            # Bijhouden welke vegetatietypen als kwalificerend zijn gerekend
+            tegengekomen_kwal_vegtypen.update(tegengekomen_kwal_vegtypen_vlak)
+
+            if bedekking_kwal_complexdelen >= self.mozaiek_minimum_bedekking:
+                omringing_kwal_vlakken += vlak_group.iloc[0].omringing_percentage
+                continue
+
+            # Als de totale bedekking kwalificerende complexdelen + HXXXX bedekking meer is dan
+            # de minimale bedekkingsthreshold, zou dit vlak in volgende ronden alsnog kunnen gaan kwalificeren
+            if (
+                bedekking_HXXXX + bedekking_kwal_complexdelen
+                > self.mozaiek_minimum_bedekking
+            ):
+                omringing_HXXXX += vlak_group.iloc[0].omringing_percentage
+                continue
+
+        return (omringing_kwal_vlakken, omringing_HXXXX, tegengekomen_kwal_vegtypen)
+
+    def _bepaal_kwalificerende_en_HXXXX_bedekking(
+        self, vlak_group: pd.DataFrame
+    ) -> Tuple[Union[int, float], Union[int, float], Set[Union[SBB, VvN]]]:
+        # Bepalen we of alle kwalificerende complexdelen samen meer dan self.mozaiek_minimum_bedekking bedekking hebben
+        bedekking_kwal_complexdelen = 0
+        bedekking_HXXXX = 0
+        tegengekomen_kwal_vegtypen = set()
+
+        for row in vlak_group.itertuples():
+            if (
+                # Als het habitattype matcht en er wordt aan de kwaliteitseisen voldaan
+                row.habtype == self.kwalificerend_habtype
+                and (row.kwaliteit == Kwaliteit.GOED or not self.alleen_goede_kwaliteit)
+            ) or (
+                # Of als de mozaiekvegetaties ook toegestaan zijn en daar is een match
+                self.ook_mozaiekvegetaties
+                and any(
```

Comment:
In theorie zijn dit allemaal set operaties.. al is dit wel lekker leesbaar zo :-)

Reply:
Is het veel sneller met set operaties? Comprehension is ook best vlot toch.